File: app/src/main/java/com/projectmanagement/adapter/in/web/ProjectController.java
package com.projectmanagement.adapter.in.web;

import com.projectmanagement.application.dto.CreateProjectRequest;
import com.projectmanagement.application.dto.ProjectResponse;
import com.projectmanagement.application.dto.UpdateProjectRequest;
import com.projectmanagement.domain.exception.ProjectNotFoundException;
import com.projectmanagement.domain.port.in.ProjectUseCase;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

/**
 * REST adapter (inbound) for project-related operations.
 *
 * <p>Translates HTTP requests into use-case calls and maps domain exceptions to
 * appropriate HTTP responses. This class must not contain any business logic —
 * it is purely a translation layer.</p>
 */
@RestController
@RequestMapping("/api/v1/projects")
public class ProjectController {

    private final ProjectUseCase projectUseCase;

    public ProjectController(ProjectUseCase projectUseCase) {
        this.projectUseCase = projectUseCase;
    }

    // -------------------------------------------------------------------------
    // POST /api/v1/projects
    // -------------------------------------------------------------------------

    /**
     * Creates a new project.
     *
     * @param request validated creation payload
     * @return 201 Created with the new project in the response body
     */
    @PostMapping
    public ResponseEntity<ProjectResponse> createProject(
            @Valid @RequestBody CreateProjectRequest request) {

        ProjectResponse response = projectUseCase.createProject(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    // -------------------------------------------------------------------------
    // GET /api/v1/projects/{id}
    // -------------------------------------------------------------------------

    /**
     * Retrieves a single project by ID.
     *
     * @param id the project UUID path variable
     * @return 200 OK with the project, or 404 Not Found if it does not exist
     */
    @GetMapping("/{id}")
    public ResponseEntity<ProjectResponse> getProjectById(@PathVariable UUID id) {
        ProjectResponse response = projectUseCase.getProjectById(id);
        return ResponseEntity.ok(response);
    }

    // -------------------------------------------------------------------------
    // GET /api/v1/projects
    // -------------------------------------------------------------------------

    /**
     * Lists all non-deleted projects.
     *
     * @return 200 OK with a (possibly empty) list of projects
     */
    @GetMapping
    public ResponseEntity<List<ProjectResponse>> listProjects() {
        List<ProjectResponse> projects = projectUseCase.listProjects();
        return ResponseEntity.ok(projects);
    }

    // -------------------------------------------------------------------------
    // PUT /api/v1/projects/{id}
    // -------------------------------------------------------------------------

    /**
     * Updates an existing project.
     *
     * @param id      the project UUID path variable
     * @param request the update payload
     * @return 200 OK with the updated project, or 404 Not Found
     */
    @PutMapping("/{id}")
    public ResponseEntity<ProjectResponse> updateProject(
            @PathVariable UUID id,
            @Valid @RequestBody UpdateProjectRequest request) {

        ProjectResponse response = projectUseCase.updateProject(id, request);
        return ResponseEntity.ok(response);
    }

    // -------------------------------------------------------------------------
    // DELETE /api/v1/projects/{id}
    // -------------------------------------------------------------------------

    /**
     * Soft-deletes a project.
     *
     * @param id the project UUID path variable
     * @return 204 No Content on success, or 404 Not Found
     */
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteProject(@PathVariable UUID id) {
        projectUseCase.deleteProject(id);
        return ResponseEntity.noContent().build();
    }

    // -------------------------------------------------------------------------
    // Exception handlers
    // -------------------------------------------------------------------------

    /**
     * Handles {@link ProjectNotFoundException} thrown by the domain service and
     * returns a 404 response with a descriptive message.
     */
    @ExceptionHandler(ProjectNotFoundException.class)
    public ResponseEntity<ErrorResponse> handleProjectNotFound(ProjectNotFoundException ex) {
        ErrorResponse error = new ErrorResponse(
                HttpStatus.NOT_FOUND.value(),
                ex.getMessage()
        );
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
    }

    /**
     * Minimal error response body.
     */
    public record ErrorResponse(int status, String message) {}
}

File: app/src/adapters/inbound/http/controllers/ProjectController.js
'use strict';

const logger = require('../../../../infrastructure/logger');

/**
 * ProjectController — inbound HTTP adapter.
 *
 * Translates HTTP requests into use-case calls and HTTP responses.
 * All business logic lives in the use-cases; this class only handles
 * HTTP concerns (parsing, status codes, response shaping).
 */
class ProjectController {
  /**
   * @param {object} usecases
   * @param {import('../../../../domain/usecases/CreateProject')} usecases.createProject
   * @param {import('../../../../domain/usecases/GetProject')}    usecases.getProject
   * @param {import('../../../../domain/usecases/ListProjects')}  usecases.listProjects
   * @param {import('../../../../domain/usecases/UpdateProject')} usecases.updateProject
   * @param {import('../../../../domain/usecases/DeleteProject')} usecases.deleteProject
   */
  constructor({ createProject, getProject, listProjects, updateProject, deleteProject }) {
    this.createProject = createProject;
    this.getProject = getProject;
    this.listProjects = listProjects;
    this.updateProject = updateProject;
    this.deleteProject = deleteProject;
  }

  /**
   * GET /api/v1/projects
   *
   * @param {import('express').Request} req
   * @param {import('express').Response} res
   * @param {import('express').NextFunction} next
   */
  async list(req, res, next) {
    try {
      const filters = {};
      if (req.query.ownerId) filters.ownerId = req.query.ownerId;
      if (req.query.status) filters.status = req.query.status;

      const projects = await this.listProjects.execute(filters);
      res.status(200).json({ data: projects });
    } catch (err) {
      next(err);
    }
  }

  /**
   * POST /api/v1/projects
   *
   * @param {import('express').Request} req
   * @param {import('express').Response} res
   * @param {import('express').NextFunction} next
   */
  async create(req, res, next) {
    try {
      const project = await this.createProject.execute(req.body);
      logger.info('Project created', { projectId: project.id, userId: req.user?.sub });
      res.status(201).json({ data: project });
    } catch (err) {
      next(err);
    }
  }

  /**
   * GET /api/v1/projects/:id
   *
   * @param {import('express').Request} req
   * @param {import('express').Response} res
   * @param {import('express').NextFunction} next
   */
  async getById(req, res, next) {
    try {
      const project = await this.getProject.execute(req.params.id);
      res.status(200).json({ data: project });
    } catch (err) {
      next(err);
    }
  }

  /**
   * PUT /api/v1/projects/:id
   *
   * @param {import('express').Request} req
   * @param {import('express').Response} res
   * @param {import('express').NextFunction} next
   */
  async update(req, res, next) {
    try {
      const project = await this.updateProject.execute(req.params.id, req.body);
      logger.info('Project updated', { projectId: project.id, userId: req.user?.sub });
      res.status(200).json({ data: project });
    } catch (err) {
      next(err);
    }
  }

  /**
   * DELETE /api/v1/projects/:id
   *
   * @param {import('express').Request} req
   * @param {import('express').Response} res
   * @param {import('express').NextFunction} next
   */
  async remove(req, res, next) {
    try {
      await this.deleteProject.execute(req.params.id);
      logger.info('Project deleted', { projectId: req.params.id, userId: req.user?.sub });
      res.status(204).send();
    } catch (err) {
      next(err);
    }
  }
}

module.exports = ProjectController;
